draft-ietf-quic-recovery.md

```diff
@@ -867,26 +867,41 @@ Pseudocode for DetectLostPackets follows:
 DetectLostPackets(largest_acked):
   loss_time = 0
   lost_packets = {}
-  delay_until_lost = kTimeThreshold *
-                     max(latest_rtt, smoothed_rtt)
-  foreach (unacked < largest_acked.packet_number):
-    time_since_sent = now() - unacked.time_sent
-    delta = largest_acked.packet_number - unacked.packet_number
-    if (time_since_sent > delay_until_lost ||
-        delta > kPacketThreshold):
+  loss_delay = kTimeThreshold * max(latest_rtt, smoothed_rtt)
+
+  // Packets sent before this time are deemed lost.
+  lost_send_time = now() - loss_delay
+
+  // Packets with packet numbers before this are deemed lost.
+  lost_pn = largest_acked.packet_number - kPacketThreshold
+
+  foreach unacked in sent_packets:
+    if (unacked.packet_number > largest_acked.packet_number):
+      continue
+
+    // Mark packet as lost, or set time when it should be marked.
+    if (unacked.time_sent <= lost_send_time ||
+        unacked.packet_number <= lost_pn):
       sent_packets.remove(unacked.packet_number)
       if (unacked.retransmittable):
         lost_packets.insert(unacked)
-    else if (loss_time == 0 && delay_until_lost != infinite):
-      loss_time = now() + delay_until_lost - time_since_sent
+    else if (loss_time == 0):
+      loss_time = unacked.time_sent + loss_delay
+    else:
+      loss_time = min(loss_time, unacked.time_sent + loss_delay)
 
   // Inform the congestion controller of lost packets and
-  // lets it decide whether to retransmit immediately.
+  // let it decide whether to retransmit immediately.
   if (!lost_packets.empty()):
     OnPacketsLost(lost_packets)
 ~~~
 
+This algorithm results in loss_time being set based on the earliest packet that
+is still in flight.  Timers set based on a loss_time that has already passed
```

Comment:
I think this comment about timers with timeouts that have already passed is more widely applicable, so maybe move it closer to SetLossDetectionTimer()?